services/src/main/java/org/keycloak/utils/ReservedCharValidator.java

```diff
@@ -19,29 +19,46 @@
 import javax.ws.rs.BadRequestException;
 import org.jboss.logging.Logger;
 
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
 /**
  *
  * @author Stan Silvert
+ * @author Lukas Hanusovsky lhanusov@redhat.com
  */
 public class ReservedCharValidator {
     protected static final Logger logger = Logger.getLogger(ReservedCharValidator.class);
     
     // https://tools.ietf.org/html/rfc3986#section-2.2
-    private static final String[] RESERVED_CHARS = { ":", "/", "?", "#", "[", "@", "!", "$", 
-                                                   "&", "(", ")", "*", "+", ",", ";", "=", 
-                                                   "]", "[", "\\" };
+    private static final Pattern RESERVED_CHARS_PATTERN = Pattern.compile("[:/?#@!$&()*+,;=\\[\\]\\\\]");
+
+    // KEYCLOAK-14231 - Supported Locales: Three new characters were added on top of this RFC: "{", "}", "%"
+    private static final Pattern RESERVED_CHARS_LOCALES_PATTERN = Pattern.compile("[:/?#@!$&()*+,;=\\[\\]\\\\{}%]");
+
     private ReservedCharValidator() {}
-    
-    public static void validate(String str) throws ReservedCharException {
+
+    public static void validate(String str, Pattern pattern) throws ReservedCharException {
+        Matcher matcher = pattern.matcher(str);
+        if (matcher.find()) {
+            String message = "Character '" + matcher.group() + "' not allowed.";
+            ReservedCharException e = new ReservedCharException(message);
+            logger.warn(message, e);
+            throw e;
```

Comment:
This is weird and I understand that this is based on the original code. Please replace with
```suggestion
            logger.warn(message);
            throw new ReservedCharException(message);
```